File: app/src/main/java/com/afkanerd/deku/Router/Router/RouterWorkManager.java
package com.afkanerd.deku.Router.Router;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.work.Worker;
import androidx.work.WorkerParameters;

import com.android.volley.ParseError;
import com.android.volley.ServerError;

import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeoutException;

public class RouterWorkManager extends Worker {
    public static final String SMS_JSON_OBJECT = "SMS_JSON_OBJECT";
    public static final String SMS_JSON_ROUTING_URL = "SMS_JSON_ROUTING_URL";
    // https://developer.android.com/topic/libraries/architecture/workmanager/basics
    public RouterWorkManager(@NonNull Context context, @NonNull WorkerParameters workerParams) {
        super(context, workerParams);
    }

    @NonNull
    @Override
    public Result doWork() {
        try {
            String gatewayServerUrl = getInputData().getString(SMS_JSON_ROUTING_URL);
            String jsonBody = getInputData().getString(SMS_JSON_OBJECT);

            if(jsonBody != null)
                RouterHandler.routeJsonMessages(getApplicationContext(), jsonBody, gatewayServerUrl);
        } catch (ExecutionException | TimeoutException | InterruptedException e) {
            e.printStackTrace();
            Throwable cause = e.getCause();
            if (cause instanceof ServerError) {
                ServerError error = (ServerError) cause;
                int statusCode = error.networkResponse.statusCode;
                if (statusCode >= 400)
                    return Result.failure();
            }
            else if(cause instanceof ParseError) {
                return Result.success();
            }
            return Result.retry();
        } catch (Exception e ) {
            e.printStackTrace();
            return Result.failure();
        }
        return Result.success();
    }
}

File: app/src/main/java/com/afkanerd/deku/Router/Models/RouterWorkManager.kt
package com.afkanerd.deku.Router.Models

import android.content.Context
import android.util.Log
import androidx.work.Worker
import androidx.work.WorkerParameters
import com.afkanerd.deku.Datastore
import com.afkanerd.deku.Modules.Network
import com.afkanerd.deku.Router.FTP
import com.afkanerd.deku.Router.SMTP
import com.sun.mail.util.MailConnectException
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import javax.mail.MessagingException
import javax.mail.SendFailedException

class RouterWorkManager (context: Context, workerParams: WorkerParameters)
    : Worker(context, workerParams) {
    override fun doWork(): Result {
        val gatewayServerId = inputData.getLong(GATEWAY_SERVER_ID, -1)
        val conversationId = inputData.getString(CONVERSATION_ID)

        val datastore = Datastore.getDatastore(applicationContext)
        val gatewayServer = datastore.gatewayServerDAO()[gatewayServerId.toString()]
        val conversation = datastore.conversationDao().getMessage(conversationId)

        val routerItem = RouterItem(conversation)
        routerItem.tag = gatewayServer.getTag()

        val jsonStringBody = routerItem.serializeJson()
        println(jsonStringBody)

        when(gatewayServer.getProtocol()) {
            SMTP.PROTOCOL -> {
                try {
                    RouterHandler.routeSmtpMessages(jsonStringBody, gatewayServer)
                } catch (e: Exception) {
                    e.printStackTrace()
                    if (e is MailConnectException) { return Result.retry() }
                    return Result.failure()
                }
            }
            FTP.PROTOCOL -> {
                try {
                    RouterHandler.routeFTPMessages(jsonStringBody, gatewayServer)
                } catch (e: Exception) {
                    Log.e(javaClass.getName(), "Exception: ", e)
                    return Result.failure()
                }
            }
            else -> {
                try {
                    when(Network.jsonRequestPost(gatewayServer.url, jsonStringBody)
                            .response.statusCode) {
                        in 500..600 -> Result.retry()
                        else -> Result.failure()
                    }
                } catch(e: Exception) {
                    Log.e(javaClass.name, "Exception routing", e)
                    Result.retry()
                }
            }
        }

        return Result.success()
    }

    companion object {
        var GATEWAY_SERVER_ID = "GATEWAY_SERVER_ID"
        var CONVERSATION_ID = "CONVERSATION_ID"
    }
}
